client/src/unifyfs-fixed.c

```diff
@@ -471,6 +472,96 @@ static int unifyfs_split_index(
     return UNIFYFS_SUCCESS;
 }
 
+/*
+ * Clear all entries in the log index.  This only clears the metadata,
+ * not the data itself.
+ */
+static void
+unifyfs_clear_index(void)
+{
+    off_t num_entries = *unifyfs_indices.ptr_num_entries;
+    size_t index_size = sizeof(*unifyfs_indices.index_entry) * num_entries;
+
+    memset(unifyfs_indices.index_entry, 0, index_size);
+    *unifyfs_indices.ptr_num_entries = 0;
+}
+
+/* Add the metadata for a single write to the index */
+static int unifyfs_logio_add_write_meta_to_index(int gfid, off_t file_pos,
+    off_t log_pos, size_t length)
+{
+    /* define an new index entry for this write operation */
+    unifyfs_index_t cur_idx;
+    cur_idx.gfid     = gfid;
+    cur_idx.file_pos = file_pos;
+    cur_idx.log_pos  = log_pos;
+    cur_idx.length   = length;
+
+    /* lookup number of existing index entries */
+    off_t num_entries = *(unifyfs_indices.ptr_num_entries);
+
+    /* get pointer to index array */
+    unifyfs_index_t* idxs = unifyfs_indices.index_entry;
+
+    /* attempt to coalesce contiguous index entries if we
+     * have an existing index in the buffer */
+    if (num_entries > 0) {
+        /* get pointer to last element in index array */
+        unifyfs_index_t* prev_idx = &idxs[num_entries - 1];
+
+        /* attempt to coalesce current index with last index,
+         * updates fields in last index and current index
+         * accordingly */
+        unifyfs_coalesce_index(prev_idx, &cur_idx,
+            unifyfs_key_slice_range);
+    }
+
+    /* add new index entries if needed */
+    while (cur_idx.length > 0) {
+        /* remaining entries we can fit in the shared memory region */
+        off_t remaining_entries = unifyfs_max_index_entries - num_entries;
+
+        /* if we have filled the key/value buffer, flush it to server */
+        if (0 == remaining_entries) {
+            /* index buffer is full, flush it */
+            int ret = invoke_client_fsync_rpc(cur_idx.gfid);
```

Comment:
There are a couple places like this where we can invoke the fsync rpc as the user is writing.  There is another one in unifyfs_split_index.  This is when the user has filled up the metadata index.  I suppose we should also rewrite and/or blank out the segment tree at these points.  Do you do that now?

Reply:
You bring up a really good issue.  I don't think things will work with my patch with that sync in there.  There may be a way I can reorder things so that the potential "sync on full index"  is done before the seg_tree_add.  Let me look into it and update the patch.